### DISCUSSION TITLE
Precision Recall Curve

### DISCUSSION THREAD

**Discussion by IssaIssa1** [ultralytics/yolov5#1994] [Q&A] [answered]
Hello I would like to ask the on what dataset does Precision Recall Curve get generated in the result folder? (train, valid and/or test)

**Comment by glenn-jocher** [comment, marked as answer]
@IssaIssa1 precision recall curve is generated from validation set after training completes.

**Reply by alevangel** [reply]
How to generate same graphs for a new test dataset?

**Reply by saribolagh** [reply]
@alevangel Just set the right parameters in the val.py and it will generate all outputs that you may need. 
I added these line to log the accuracy and confusion matrix as well.
    if save_stats: 
        os.makedirs(str(save_dir / 'stats'), exist_ok=True)
        with open(str(save_dir / 'stats' /'accuracy.csv'), 'w') as fh:
            fh.write(f'Class,Images,Labels,Precision,Recall,mAP@.5,mAP@.5:.95,conf_thres,iou_thres,task\n')
            fh.write(f"all, {seen}, {nt.sum()}, {mp}, {mr}, {map50}, {map},{conf_thres},{iou_thres},{task}\n")
            for i, c in enumerate(ap_class):
                fh.write(f'{names[c]}, {seen}, {nt[c]}, {p[i]}, {r[i]}, {ap50[i]}, {ap[i]}, {conf_thres},{iou_thres},{task}\n')
        
        confusion_matrix.export(str(save_dir / 'stats'/'cm.csv'), list(names.values()))

**Reply by glenn-jocher** [reply]
@alevangel @saribolagh val.py can run on the test set with `--task test` (default is `--task val`):
```
python val.py --data DATA.yaml --task test
```

**Reply by b0gfasdfg** [reply]
can I know what do the label correlogram mean?

**Reply by glenn-jocher** [reply]
@b0gfasdfg correlogram simply displays relationships among label dimensions xywh.

**Reply by JordanWei123** [reply]
@glenn-jocher How to export PR curve data to excel file

**Comment by kashishgoyal31** [comment]
how do I analyse the result.png file after training
Pls expalin the parameters

**Reply by glenn-jocher** [reply]
@kashishgoyal31 `results.png` displays train and val losses and metrics (P, R, mAP)

![image](https://user-images.githubusercontent.com/26833433/129556555-237f3f7e-79ae-4875-83d7-5201415e71ea.png)

**Reply by kashishgoyal31** [reply]
How to decide the no. of epochs when we train the model. Is there some specification w.r.t to the size of dataset or is there some other parameter that decides the epochs

**Reply by glenn-jocher** [reply]
@kashishgoyal31 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Reply by kashishgoyal31** [reply]
How can I improve the code to get the better results? What changes shall be made to tune the parameters to get more precise results

**Reply by Artikaa19** [reply]
Hello sir, how i can show the recall and precision like you before show the result.png?

**Reply by glenn-jocher** [reply]
@Artikaa19 results.png is plotted from results.csv upon training completion. You can also plot any results.csv manually with `plot_results()`:
https://github.com/ultralytics/yolov5/blob/1156a323f8fc221d5c1a36c700e72e129419d187/utils/plots.py#L417-L418

**Comment by kashishgoyal31** [comment]
Dear Sir,

What is the difference between ***@***.******@***.***:.95? Which shall be considered
for model evaluation?


Regards
Kashish Goyal

**Reply by glenn-jocher** [reply]
@kashishgoyal31 they're both variously considered by different competitions, i.e. VOC uses mAP@0.5, COCO uses @0.5:0.95.

**Comment by kashishgoyal31** [comment]
Which shall we consider for our model evaluation as map value??


Regards
Kashish Goyal

**Reply by glenn-jocher** [reply]
@kashishgoyal31 **mAP@0.5:0.95** is the generally accepted object detection metric I would say.

**Reply by alevangel** [reply]
@kashishgoyal31 just for reference: https://datascience.stackexchange.com/questions/16797/what-does-the-notation-map-5-95-mean

**Comment by b0gfasdfg** [comment]
thanks




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ultralytics/yolov5"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年5月18日(星期三) 晚上9:42
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ultralytics/yolov5] Precision Recall Curve (Discussion #1994)





 
@b0gfasdfg correlogram simply displays relationships among label dimensions xywh.

**Comment by farhanhumayun** [comment]
Hi @glenn-jocher, I want to manually plot the P-R and F1-Confidence curves for comparative analysis of various models. How can I do that from the information in results.csv?

**Reply by HsinChinLi** [reply]
Hi @farhanhumayun , I believe that you can find the parameters and print/save them from metrics.py. For instance, adding print commands in metrics.py and ran val.py to print out the lists of Precision and Recall.

**Reply by beneon** [reply]
And I found that in yolo/utils/metrics.py 487:492, there are multiple references to plot_pr_curve and plot_mc_curve, using px, py, p, r and other things:

```{python}
    if plot:
        plot_pr_curve(px, py, ap, save_dir / f'{prefix}PR_curve.png', names, on_plot=on_plot)
        plot_mc_curve(px, f1, save_dir / f'{prefix}F1_curve.png', names, ylabel='F1', on_plot=on_plot)
        plot_mc_curve(px, p, save_dir / f'{prefix}P_curve.png', names, ylabel='Precision', on_plot=on_plot)
        plot_mc_curve(px, r, save_dir / f'{prefix}R_curve.png', names, ylabel='Recall', on_plot=on_plot)

```

So maybe we can export these data to files at this point.
